crates/ra_hir/src/from_source.rs

```diff
@@ -0,0 +1,217 @@
+use ra_db::{FileId, FilePosition};
+use ra_syntax::{
+    algo::find_node_at_offset,
+    ast::{self, AstNode, NameOwner},
+    SyntaxNode,
+};
+
+use crate::{
+    db::{AstDatabase, DefDatabase, HirDatabase},
+    ids::{AstItemDef, ConstId, EnumId, FunctionId, StaticId, StructId, TraitId, TypeAliasId},
+    name::AsName,
+    Const, Crate, Enum, EnumVariant, FieldSource, Function, HasSource, ImplBlock, Module,
+    ModuleSource, Source, Static, Struct, StructField, Trait, TypeAlias, Union, VariantDef,
+};
+
+pub trait FromSource {
+    type Ast;
+    fn from_source(db: &(impl DefDatabase + AstDatabase), src: Source<Self::Ast>) -> Option<Self>
+    where
+        Self: Sized;
```

Comment:
We don't have any object-safe methods, so it's best to move `: Sized` bound to the trait itself: `trait FromSource: Sized`